File: scripts/reddit-scraper.js
// Reddit scraper — mines cold-outreach wisdom from a set of subreddits.
// Read-only, polite (rate-limited, resumable). Zero npm install (uses Node fetch).
//
// SETUP (one time): create a Reddit app at https://www.reddit.com/prefs/apps
//   -> "create app" -> type "script" -> note the client id (under the name) + secret.
// Then set these as system env vars (per your key preference):
//   export REDDIT_CLIENT_ID="..."      # required
//   export REDDIT_CLIENT_SECRET="..."  # required for a script/web app
//   export REDDIT_USERNAME="..."       # optional (enables the password grant)
//   export REDDIT_PASSWORD="..."       # optional
// RUN:
//   zsh -ic 'node scripts/reddit-scraper.js'            # scrape all subs
//   zsh -ic 'node scripts/reddit-scraper.js sales'      # just one sub
// Output: data/reddit/corpus.json  (array of {subreddit, id, title, url, ups, text})
import { mkdirSync, readFileSync, writeFileSync, existsSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';

const __dirname = dirname(fileURLToPath(import.meta.url));
const OUT_DIR = join(__dirname, '..', 'data', 'reddit');
const CORPUS = join(OUT_DIR, 'corpus.json');
mkdirSync(OUT_DIR, { recursive: true });

const UA = 'wahpaki-crm:reddit-scraper:0.1 (outreach research; by /u/andrew)';
const SUBREDDITS = process.argv.slice(2).filter((a) => !a.startsWith('-'));
const SUBS = SUBREDDITS.length ? SUBREDDITS
  : ['sales', 'salestechniques', 'Entrepreneur', 'freelancing', 'LeadGeneration', 'ideavalidation', 'PromptEngineering', 'ChatGPTPro', 'SideProject'];
const QUERIES = ['cold email', 'cold outreach', 'cold email template', 'cold email reply rate', 'sales email', 'cold email prompt', 'follow up sequence', 'email deliverability'];
const POSTS_PER_QUERY = 15;
const COMMENTS_PER_POST = 60;
const MIN_TEXT = 40; // skip trivially short comments/posts

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

async function getToken() {
  const id = process.env.REDDIT_CLIENT_ID, secret = process.env.REDDIT_CLIENT_SECRET || '';
  if (!id) { const e = new Error('REDDIT_CLIENT_ID is not set. Create a Reddit "script" app and export the id/secret. See the header of this file.'); e.code = 'NO_KEY'; throw e; }
  const basic = Buffer.from(`${id}:${secret}`).toString('base64');
  const body = new URLSearchParams();
  if (process.env.REDDIT_USERNAME && process.env.REDDIT_PASSWORD) {
    body.set('grant_type', 'password');
    body.set('username', process.env.REDDIT_USERNAME);
    body.set('password', process.env.REDDIT_PASSWORD);
  } else if (secret) {
    body.set('grant_type', 'client_credentials');
  } else {
    body.set('grant_type', 'https://oauth.reddit.com/grants/installed_client');
    body.set('device_id', 'DO_NOT_TRACK_THIS_DEVICE');
  }
  const res = await fetch('https://www.reddit.com/api/v1/access_token', {
    method: 'POST', headers: { Authorization: `Basic ${basic}`, 'Content-Type': 'application/x-www-form-urlencoded', 'User-Agent': UA }, body,
  });
  const j = await res.json().catch(() => ({}));
  if (!res.ok || !j.access_token) throw new Error(`Reddit auth failed (${res.status}): ${j.error || JSON.stringify(j).slice(0, 200)}`);
  return j.access_token;
}

let ACCESS = null;
async function token() { if (!ACCESS) ACCESS = await getToken(); return ACCESS; }

async function api(path, tries = 3) {
  for (let i = 0; i < tries; i++) {
    const res = await fetch(`https://oauth.reddit.com${path}`, { headers: { Authorization: `bearer ${await token()}`, 'User-Agent': UA } });
    if (res.status === 429) { await sleep(2000 * (i + 1)); continue; }
    if (res.status === 401) { ACCESS = null; continue; } // token expired mid-run, refresh + retry
    // stay under Reddit's ~100 req/min: pace off the remaining-header when present
    const remain = Number(res.headers.get('x-ratelimit-remaining'));
    if (!Number.isNaN(remain) && remain < 3) await sleep(3000);
    if (!res.ok) throw new Error(`GET ${path} -> ${res.status}`);
    return res.json();
  }
  throw new Error(`GET ${path} failed after ${tries} tries`);
}

const clean = (s) => (s || '').replace(/\s+/g, ' ').trim();

async function run() {
  await token(); // validate creds up front so auth errors surface immediately
  const corpus = existsSync(CORPUS) ? JSON.parse(readFileSync(CORPUS, 'utf8')) : [];
  const seen = new Set(corpus.map((c) => c.id));
  let added = 0;

  for (const sub of SUBS) {
    const postIds = new Set();
    for (const q of QUERIES) {
      let listing;
      try { listing = await api(`/r/${sub}/search?q=${encodeURIComponent(q)}&restrict_sr=1&sort=top&t=all&limit=${POSTS_PER_QUERY}`); }
      catch (e) { console.log(`  ${sub} "${q}": ${e.message}`); continue; }
      for (const ch of listing?.data?.children || []) postIds.add(ch.data.id);
      await sleep(1200);
    }
    console.log(`r/${sub}: ${postIds.size} unique posts across ${QUERIES.length} queries`);

    for (const pid of postIds) {
      if (seen.has(pid)) continue;
      let thread;
      try { thread = await api(`/r/${sub}/comments/${pid}?limit=${COMMENTS_PER_POST}&depth=3&sort=top`); }
      catch (e) { console.log(`    post ${pid}: ${e.message}`); await sleep(1200); continue; }
      const post = thread?.[0]?.data?.children?.[0]?.data;
      if (!post) { await sleep(1000); continue; }
      const comments = [];
      const walk = (nodes) => { for (const n of nodes || []) { if (n.kind === 't1' && n.data?.body) { const b = clean(n.data.body); if (b.length >= MIN_TEXT && b !== '[deleted]' && b !== '[removed]') comments.push(b); } if (n.data?.replies?.data?.children) walk(n.data.replies.data.children); } };
      walk(thread?.[1]?.data?.children);
      const text = [clean(post.title), clean(post.selftext), ...comments].filter((t) => t.length >= MIN_TEXT).join('\n\n');
      corpus.push({ subreddit: sub, id: pid, title: clean(post.title), url: `https://reddit.com${post.permalink}`, ups: post.ups, num_comments: post.num_comments, text });
      seen.add(pid); added++;
      writeFileSync(CORPUS, JSON.stringify(corpus, null, 2)); // trickle-save so it's resumable
      process.stdout.write(`\r    saved ${added} new threads…`);
      await sleep(1400);
    }
    process.stdout.write('\n');
  }
  const bySub = {};
  for (const c of corpus) bySub[c.subreddit] = (bySub[c.subreddit] || 0) + 1;
  console.log('Per subreddit:', Object.entries(bySub).map(([s, n]) => `${s}:${n}`).join('  '));
  console.log(`\nDone. ${corpus.length} threads in ${CORPUS} (${added} new this run).`);
}

run().catch((e) => { console.error(e.message); process.exit(1); });

File: scripts/reddit-scraper.py
#!/usr/bin/env python3
"""Collect public Reddit threads using the current Chrome Reddit session.

The scraper is read-only, rate-limited, and resumable. It stores no Reddit
cookie or account information; only public post/comment text is written to
data/reddit/corpus.json.

Examples:
    python3 scripts/reddit-scraper.py
    python3 scripts/reddit-scraper.py sales LeadGeneration
    python3 scripts/reddit-scraper.py sales --max-new 5 --delay 0.5
"""

import argparse
import json
import os
import re
import sys
import time
from pathlib import Path
from urllib.parse import quote

try:
    import browser_cookie3
    import requests
except ImportError as exc:
    missing = getattr(exc, "name", "a required package")
    raise SystemExit(
        f"Missing {missing}. Run: python3 -m pip install -r requirements-reddit.txt"
    ) from exc


ROOT = Path(__file__).resolve().parent.parent
# The corpus path is overridable so a focused variant (for example the
# second-touchpoint pipeline) keeps its own corpus. Relative paths resolve
# against the repo root.
_reddit_corpus_override = os.environ.get("WAHPAKI_REDDIT_CORPUS", "").strip()
CORPUS = (
    (Path(_reddit_corpus_override) if Path(_reddit_corpus_override).is_absolute() else ROOT / _reddit_corpus_override)
    if _reddit_corpus_override
    else ROOT / "data" / "reddit" / "corpus.json"
)
OUT_DIR = CORPUS.parent
USER_AGENT = (
    "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) "
    "AppleWebKit/537.36 (KHTML, like Gecko) "
    "Chrome/126.0 Safari/537.36"
)
def _env_list(name, default):
    """Split a comma/newline-separated env override into a clean list."""
    raw = os.environ.get(name, "")
    values = [part.strip() for part in re.split(r"[\n,]+", raw) if part.strip()]
    return values or default


DEFAULT_SUBREDDITS = _env_list(
    "WAHPAKI_REDDIT_SUBREDDITS",
    [
        "sales",
        "salestechniques",
        "Entrepreneur",
        "freelancing",
        "LeadGeneration",
        "ideavalidation",
        "PromptEngineering",
        "ChatGPTPro",
        "SideProject",
    ],
)
QUERIES = _env_list(
    "WAHPAKI_REDDIT_QUERIES",
    [
        "cold email",
        "cold outreach",
        "cold email template",
        "cold email reply rate",
        "sales email",
        "cold email prompt",
        "follow up sequence",
        "email deliverability",
    ],
)
MIN_TEXT = 40


def parse_args():
    parser = argparse.ArgumentParser(
        description="Build a public Reddit cold-outreach corpus using Chrome cookies."
    )
    parser.add_argument(
        "subreddits",
        nargs="*",
        help="Subreddits to scrape (default: the full configured list).",
    )
    parser.add_argument(
        "--posts-per-query",
        type=int,
        default=10,
        metavar="N",
        help="Search results requested per query (default: 10).",
    )
    parser.add_argument(
        "--comments-per-post",
        type=int,
        default=60,
        metavar="N",
        help="Top comments requested per thread (default: 60).",
    )
    parser.add_argument(
        "--max-new",
        type=int,
        default=0,
        metavar="N",
        help="Stop after N newly saved threads; 0 means no limit.",
    )
    parser.add_argument(
        "--max-new-per-subreddit",
        type=int,
        default=0,
        metavar="N",
        help="Save at most N new threads from each subreddit; 0 means no limit.",
    )
    parser.add_argument(
        "--delay",
        type=float,
        default=1.5,
        metavar="SECONDS",
        help="Delay between Reddit requests (default: 1.5).",
    )
    args = parser.parse_args()
    if args.posts_per_query < 1 or args.comments_per_post < 1:
        parser.error("request limits must be positive integers")
    if args.max_new < 0 or args.max_new_per_subreddit < 0 or args.delay < 0:
        parser.error("--max-new, --max-new-per-subreddit, and --delay cannot be negative")
    args.subreddits = [
        re.sub(r"^r/", "", subreddit.strip(), flags=re.IGNORECASE)
        for subreddit in (args.subreddits or DEFAULT_SUBREDDITS)
        if subreddit.strip()
    ]
    return args


def clean(value):
    return re.sub(r"\s+", " ", value or "").strip()


def load_corpus():
    if not CORPUS.exists():
        return []
    try:
        corpus = json.loads(CORPUS.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as exc:
        raise SystemExit(f"Cannot read existing corpus {CORPUS}: {exc}") from exc
    if not isinstance(corpus, list):
        raise SystemExit(f"Existing corpus must be a JSON array: {CORPUS}")
    return corpus


def save_corpus(corpus):
    """Atomically replace the corpus so an interruption cannot corrupt it."""
    OUT_DIR.mkdir(parents=True, exist_ok=True)
    temporary = CORPUS.with_suffix(".json.tmp")
    temporary.write_text(
        json.dumps(corpus, indent=2, ensure_ascii=False) + "\n",
        encoding="utf-8",
    )
    os.replace(temporary, CORPUS)


def walk_comments(nodes, output):
    for node in nodes or []:
        data = node.get("data") or {}
        if node.get("kind") == "t1":
            body = clean(data.get("body"))
            if len(body) >= MIN_TEXT and body not in ("[deleted]", "[removed]"):
                output.append(body)
        replies = data.get("replies")
        if isinstance(replies, dict):
            walk_comments((replies.get("data") or {}).get("children"), output)


class RedditClient:
    def __init__(self, delay):
        try:
            cookie_jar = browser_cookie3.chrome(domain_name="reddit.com")
        except Exception as exc:
            raise SystemExit(
                "Could not read Reddit cookies from Chrome. Make sure Chrome has a "
                "Reddit session and approve the macOS Keychain prompt, then retry. "
                f"Details: {exc}"
            ) from exc
        if not list(cookie_jar):
            raise SystemExit(
                "No reddit.com cookies were found in Chrome. Log in to Reddit in "
                "Chrome, then retry."
            )
        self.delay = delay
        self.session = requests.Session()
        self.session.cookies.update(cookie_jar)
        self.session.headers.update(
            {"User-Agent": USER_AGENT, "Accept": "application/json"}
        )
        print(f"Loaded {len(cookie_jar)} reddit.com cookies from Chrome.")

    def get_json(self, url, *, params=None, tries=3):
        last_error = None
        for attempt in range(1, tries + 1):
            try:
                response = self.session.get(
                    url, params=params, timeout=(10, 30)
                )
                if response.status_code == 429:
                    retry_after = response.headers.get("retry-after")
                    wait = float(retry_after) if retry_after else 3 * attempt
                    last_error = RuntimeError(
                        f"Reddit rate limit (HTTP 429; last backoff {wait:g}s)"
                    )
                    print(f"    Reddit rate limit; waiting {wait:g}s…")
                    time.sleep(wait)
                    continue
                response.raise_for_status()
                content_type = response.headers.get("content-type", "")
                if "json" not in content_type.lower():
                    raise RuntimeError(
                        f"expected JSON, received {content_type or 'unknown content'}"
                    )
                payload = response.json()
                if self.delay:
                    time.sleep(self.delay)
                return payload
            except (requests.RequestException, ValueError, RuntimeError) as exc:
                last_error = exc
                if attempt < tries:
                    time.sleep(2 * attempt)
        raise RuntimeError(f"GET {url} failed after {tries} tries: {last_error}")


def run():
    args = parse_args()
    OUT_DIR.mkdir(parents=True, exist_ok=True)
    corpus = load_corpus()
    seen = {
        item.get("id")
        for item in corpus
        if isinstance(item, dict) and item.get("id")
    }
    client = RedditClient(args.delay)
    added = 0
    stop = False

    for subreddit in args.subreddits:
        added_for_subreddit = 0
        post_ids = []
        for query in QUERIES:
            try:
                listing = client.get_json(
                    f"https://www.reddit.com/r/{quote(subreddit, safe='')}/search.json",
                    params={
                        "q": query,
                        "restrict_sr": 1,
                        "sort": "top",
                        "t": "all",
                        "limit": args.posts_per_query,
                        "raw_json": 1,
                    },
                )
            except Exception as exc:
                print(f'  r/{subreddit} "{query}": {exc}')
                continue
            children = (listing.get("data") or {}).get("children") or []
            post_ids.extend(
                child.get("data", {}).get("id")
                for child in children
                if child.get("data", {}).get("id")
            )

        post_ids = list(dict.fromkeys(post_ids))
        unseen = [post_id for post_id in post_ids if post_id not in seen]
        print(
            f"r/{subreddit}: {len(post_ids)} unique results; "
            f"{len(unseen)} new threads"
        )

        for post_id in unseen:
            try:
                thread = client.get_json(
                    (
                        f"https://www.reddit.com/r/{quote(subreddit, safe='')}/"
                        f"comments/{quote(post_id, safe='')}.json"
                    ),
                    params={
                        "limit": args.comments_per_post,
                        "depth": 3,
                        "sort": "top",
                        "raw_json": 1,
                    },
                )
                post = thread[0]["data"]["children"][0]["data"]
            except Exception as exc:
                print(f"    {post_id}: {exc}")
                continue

            comments = []
            walk_comments(thread[1]["data"]["children"], comments)
            pieces = [
                clean(post.get("title")),
                clean(post.get("selftext")),
                *comments,
            ]
            text = "\n\n".join(piece for piece in pieces if len(piece) >= MIN_TEXT)
            if not text:
                print(f"    {post_id}: skipped (no substantive text)")
                seen.add(post_id)
                continue

            corpus.append(
                {
                    "subreddit": subreddit,
                    "id": post_id,
                    "title": clean(post.get("title")),
                    "url": "https://reddit.com" + post.get("permalink", ""),
                    "ups": post.get("ups"),
                    "num_comments": post.get("num_comments"),
                    "text": text,
                }
            )
            seen.add(post_id)
            added += 1
            added_for_subreddit += 1
            save_corpus(corpus)
            print(f"    saved {added} new thread{'s' if added != 1 else ''}")

            if args.max_new and added >= args.max_new:
                stop = True
                break
            if (
                args.max_new_per_subreddit
                and added_for_subreddit >= args.max_new_per_subreddit
            ):
                break
        if stop:
            break

    by_subreddit = {}
    for item in corpus:
        subreddit = item.get("subreddit", "unknown")
        by_subreddit[subreddit] = by_subreddit.get(subreddit, 0) + 1
    summary = "  ".join(
        f"{subreddit}:{count}" for subreddit, count in by_subreddit.items()
    )
    print(f"Per subreddit: {summary or 'none'}")
    print(f"Done. {len(corpus)} threads in {CORPUS} ({added} new this run).")


if __name__ == "__main__":
    try:
        run()
    except KeyboardInterrupt:
        print("\nStopped. Already saved threads are safe.", file=sys.stderr)
        raise SystemExit(130)
